File: external/java/src/main/java/ch/epfl/dedis/lib/crypto/bn256/TwistPoint.java
package ch.epfl.dedis.lib.crypto.bn256;

import java.math.BigInteger;

class TwistPoint {
    GFp2 x, y, z, t;

    static GFp2 twistB = new GFp2(
            new BigInteger("6500054969564660373279643874235990574282535810762300357187714502686418407178"),
            new BigInteger("45500384786952622612957507119651934019977750675336102500314001518804928850249")
    );

    static TwistPoint twistGen = new TwistPoint(
            new GFp2(
                    new BigInteger("21167961636542580255011770066570541300993051739349375019639421053990175267184"),
                    new BigInteger("64746500191241794695844075326670126197795977525365406531717464316923369116492")
            ),
            new GFp2(
                    new BigInteger("20666913350058776956210519119118544732556678129809273996262322366050359951122"),
                    new BigInteger("17778617556404439934652658462602675281523610326338642107814333856843981424549")
            ),
            new GFp2(
                    new BigInteger("0"),
                    new BigInteger("1")
            ),
            new GFp2(
                    new BigInteger("0"),
                    new BigInteger("1")
            )
    );

    TwistPoint() {
        this.x = new GFp2();
        this.y = new GFp2();
        this.z = new GFp2();
        this.t = new GFp2();
    }

    TwistPoint(TwistPoint p) {
        this.x = new GFp2(p.x);
        this.y = new GFp2(p.y);
        this.z = new GFp2(p.z);
        this.t = new GFp2(p.t);
    }

    TwistPoint(GFpPool pool) {
        x = pool.get2();
        y = pool.get2();
        z = pool.get2();
        t = pool.get2();
    }

    private TwistPoint(GFp2 x, GFp2 y, GFp2 z, GFp2 t) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.t = t;
    }

    @Override
    public String toString() {
        TwistPoint c = new TwistPoint(this);
        return "(" + c.x.toString() + "," + c.y.toString() + "," + c.z.toString() + ")";
    }

    void set(TwistPoint a) {
        this.x.set(a.x);
        this.y.set(a.y);
        this.z.set(a.z);
        this.t.set(a.t);
    }

    boolean isOnCurve() {
        GFp2 yy = GFpPool.getInstance().get2().square(this.y);
        GFp2 xxx = GFpPool.getInstance().get2().square(this.x);
        xxx.mul(xxx, this.x);
        yy.sub(yy, xxx);
        yy.sub(yy, twistB);
        yy.minimal();
        boolean result = yy.x.signum() == 0 && yy.y.signum() == 0;
        GFpPool.getInstance().put2(yy, xxx);
        return result;
    }

    void setInfinity() {
        this.z.setZero();
    }

    boolean isInfinity() {
        return this.z.isZero();
    }

    void add(TwistPoint a, TwistPoint b) {

        if (a.isInfinity()) {
            this.set(b);
            return;
        }
        if (b.isInfinity()) {
            this.set(a);
            return;
        }

        GFp2 z1z1 = GFpPool.getInstance().get2().square(a.z);
        GFp2 z2z2 = GFpPool.getInstance().get2().square(b.z);
        GFp2 u1 = GFpPool.getInstance().get2().mul(a.x, z2z2);
        GFp2 u2 = GFpPool.getInstance().get2().mul(b.x, z1z1);

        GFp2 t = GFpPool.getInstance().get2().mul(b.z, z2z2);
        GFp2 s1 = GFpPool.getInstance().get2().mul(a.y, t);

        t.mul(a.z, z1z1);
        GFp2 s2 = GFpPool.getInstance().get2().mul(b.y, t);

        GFp2 h = GFpPool.getInstance().get2().sub(u2, u1);
        GFpPool.getInstance().put2(u2);

        boolean xEqual = h.isZero();

        t.add(h, h);
        GFp2 i = GFpPool.getInstance().get2().square(t);
        GFp2 j = GFpPool.getInstance().get2().mul(h, i);

        t.sub(s2, s1);
        GFpPool.getInstance().put2(s2);

        boolean yEqual = t.isZero();
        if (xEqual && yEqual) {
            this.dbl(a);
            return;
        }
        GFp2 r = GFpPool.getInstance().get2().add(t, t);

        GFp2 v = GFpPool.getInstance().get2().mul(u1, i);

        GFpPool.getInstance().put2(i, u1);


        GFp2 t4 = GFpPool.getInstance().get2().square(r);
        t.add(v, v);
        GFp2 t6 = GFpPool.getInstance().get2().sub(t4, j);
        this.x.sub(t6, t);

        t.sub(v, this.x);
        t4.mul(s1, j);

        t6.add(t4, t4);
        t4.mul(r, t);

        this.y.sub(t4, t6);

        t.add(a.z, b.z);
        t4.square(t);
        t.sub(t4, z1z1);

        t4.sub(t, z2z2);

        this.z.mul(t4, h);

        GFpPool.getInstance().put2(s1, j, v, r, z1z1, z2z2, t, t4, t6, h);
    }

    void dbl(TwistPoint a) {
        GFp2 A = GFpPool.getInstance().get2().square(a.x);
        GFp2 B = GFpPool.getInstance().get2().square(a.y);
        GFp2 C = GFpPool.getInstance().get2().square(B);

        t = GFpPool.getInstance().get2().add(a.x, B);
        GFpPool.getInstance().put2(B);

        GFp2 t2 = GFpPool.getInstance().get2().square(t);
        t.sub(t2, A);
        t2.sub(t, C);
        GFp2 d = GFpPool.getInstance().get2().add(t2, t2);
        t.add(A, A);
        GFp2 e = GFpPool.getInstance().get2().add(t, A);
        GFpPool.getInstance().put2(A);

        GFp2 f = GFpPool.getInstance().get2().square(e);

        t.add(d, d);
        this.x.sub(f, t);

        t.add(C, C);
        t2.add(t, t);
        t.add(t2, t2);
        this.y.sub(d, this.x);
        t2.mul(e, this.y);
        this.y.sub(t2, t);

        t.mul(a.y, a.z);
        this.z.add(t, t);

        GFpPool.getInstance().put2(C, t, t2, d, e, f);
    }

    TwistPoint mul(TwistPoint a, BigInteger scalar) {
        TwistPoint sum = new TwistPoint();
        sum.setInfinity();
        TwistPoint t = new TwistPoint();

        for (int i = scalar.bitLength(); i >= 0; i--) {
            t.dbl(sum);
            if (scalar.testBit(i)) {
                sum.add(t, a);
            } else {
                sum.set(t);
            }
        }

        this.set(sum);
        return this;
    }

    TwistPoint makeAffine() {
        if (this.z.isOne()) {
            return this;
        }
        if (this.isInfinity()) {
            this.x.setZero();
            this.y.setOne();
            this.z.setZero();
            this.t.setZero();
            return this;
        }

        GFp2 zInv = GFpPool.getInstance().get2().invert(this.z);
        this.t = new GFp2().mul(this.y, zInv);
        GFp2 zInv2 = GFpPool.getInstance().get2().square(zInv);
        this.y.mul(t, zInv2);
        this.t.mul(this.x, zInv2);
        this.x.set(this.t);
        this.z.setOne();
        this.t.setOne();

        GFpPool.getInstance().put2(zInv, zInv2);

        return this;
    }

    void negative(TwistPoint a) {
        this.x.set(a.x);
        this.y.setZero();
        this.y.sub(this.y, a.y);
        this.z.set(a.z);
        this.t.setZero();
    }

    void free(GFpPool pool) {
        pool.put2(x, y, z, t);
        // Make sure any use of the object will fail.
        x = null;
        y = null;
        z = null;
        t = null;
    }
}

File: external/js/kyber/src/pairing/twist-point.ts
import BN from 'bn.js';
import GfP2 from './gfp2';

const twistB = new GfP2(
    '6500054969564660373279643874235990574282535810762300357187714502686418407178',
    '45500384786952622612957507119651934019977750675336102500314001518804928850249'
);

/**
 * Point class used by G2
 */
export default class TwistPoint {
    static generator = new TwistPoint(
        new GfP2(
            '21167961636542580255011770066570541300993051739349375019639421053990175267184',
            '64746500191241794695844075326670126197795977525365406531717464316923369116492',
        ),
        new GfP2(
            '20666913350058776956210519119118544732556678129809273996262322366050359951122',
            '17778617556404439934652658462602675281523610326338642107814333856843981424549',
        ),
        new GfP2('0', '1'),
        new GfP2('0', '1'),
    );

    private x: GfP2;
    private y: GfP2;
    private z: GfP2;
    private t: GfP2;

    constructor(x?: GfP2, y?: GfP2, z?: GfP2, t?: GfP2) {
        this.x = x || GfP2.zero();
        this.y = y || GfP2.zero();
        this.z = z || GfP2.zero();
        this.t = t || GfP2.zero();
    }

    /**
     * Get the x element of the point
     * @returns the x element
     */
    getX(): GfP2 {
        return this.x;
    }

    /**
     * Get the y element of the point
     * @returns the y element
     */
    getY(): GfP2 {
        return this.y;
    }

    /**
     * Get the z element of the point
     * @returns the z element
     */
    getZ(): GfP2 {
        return this.z;
    }

    /**
     * Get the t element of the point
     * @returns the t element
     */
    getT(): GfP2 {
        return this.t;
    }

    /**
     * Check if the point is on the curve, meaning it's a valid point
     * @returns true for a valid point, false otherwise
     */
    isOnCurve(): boolean {
        const cpy = this.clone();
        cpy.makeAffine();
        if (cpy.isInfinity()) {
            return true;
        }

        const yy = cpy.y.square();
        const xxx = cpy.x.square().mul(cpy.x).add(twistB);

        return yy.equals(xxx);
    }

    /**
     * Set the point to the infinity value
     */
    setInfinity(): void {
        this.x = GfP2.zero();
        this.y = GfP2.one();
        this.z = GfP2.zero();
        this.t = GfP2.zero();
    }

    /**
     * Check if the point is the infinity
     * @returns true when the infinity, false otherwise
     */
    isInfinity(): boolean {
        return this.z.isZero();
    }

    /**
     * Add a to b and set the value to the point
     * @param a first point
     * @param b second point
     */
    add(a: TwistPoint, b: TwistPoint): void {
        if (a.isInfinity()) {
            this.copy(b);
            return;
        }
        if (b.isInfinity()) {
            this.copy(a);
            return;
        }

        const z12 = a.z.square();
        const z22 = b.z.square();
        const u1 = a.x.mul(z22);
        const u2 = b.x.mul(z12);

        let t = b.z.mul(z22);
        const s1 = a.y.mul(t);

        t = a.z.mul(z12);
        const s2 = b.y.mul(t);

        const h = u2.sub(u1);
        
        t = h.add(h);
        const i = t.square();
        const j = h.mul(i);

        t = s2.sub(s1);
        if (h.isZero() && t.isZero()) {
            this.double(a);
            return;
        }

        const r = t.add(t);
        const v = u1.mul(i);

        let t4 = r.square();
        t = v.add(v);
        let t6 = t4.sub(j);
        this.x = t6.sub(t);

        t = v.sub(this.x);
        t4 = s1.mul(j);
        t6 = t4.add(t4);
        t4 = r.mul(t);
        this.y = t4.sub(t6);

        t = a.z.add(b.z);
        t4 = t.square();
        t = t4.sub(z12);
        t4 = t.sub(z22);
        this.z = t4.mul(h);
    }

    /**
     * Compute the double of the given point and set the value
     * @param a the point
     */
    double(a: TwistPoint): void {
        const A = a.x.square();
        const B = a.y.square();
        const C = B.square();

        let t = a.x.add(B);
        let t2 = t.square();
        t = t2.sub(A);
        t2 = t.sub(C);
        const d = t2.add(t2);
        t = A.add(A);
        const e = t.add(A);
        const f = e.square();

        t = d.add(d);
        this.x = f.sub(t);

        t = C.add(C);
        t2 = t.add(t);
        t = t2.add(t2);
        this.y = d.sub(this.x);
        t2 = e.mul(this.y);
        this.y = t2.sub(t);

        t = a.y.mul(a.z);
        this.z = t.add(t);
    }

    /**
     * Multiply a point by a scalar and set the value to the point
     * @param a the point
     * @param k the scalar
     */
    mul(a: TwistPoint, k: BN): void {
        const sum = new TwistPoint();
        const t = new TwistPoint();

        for (let i = k.bitLength(); i >= 0; i--) {
            t.double(sum);
            if (k.testn(i)) {
                sum.add(t, a);
            } else {
                sum.copy(t);
            }
        }

        this.copy(sum);
    }

    /**
     * Normalize the point coordinates
     */
    makeAffine(): void {
        if (this.z.isOne()) {
            return;
        } else if (this.z.isZero()) {
            this.setInfinity();
            return;
        }

        const zInv = this.z.invert();
        let t = this.y.mul(zInv);
        const zInv2 = zInv.square();
        this.y = t.mul(zInv2);
        this.x = this.x.mul(zInv2);
        this.z = GfP2.one();
        this.t = GfP2.one();
    }

    /**
     * Compute the negative of a and set the value to the point
     * @param a the point
     */
    neg(a: TwistPoint): void {
        this.x = a.x;
        this.y = a.y.negative();
        this.z = a.z;
    }

    /**
     * Fill the point with the values of a
     * @param a the point
     */
    copy(a: TwistPoint): void {
        this.x = a.x;
        this.y = a.y;
        this.z = a.z;
        this.t = a.t;
    }

    /**
     * Get the a clone of the current point
     * @returns a copy of the point
     */
    clone(): TwistPoint {
        return new TwistPoint(this.x, this.y, this.z, this.t);
    }

    /**
     * Check the equality between two points
     * @returns true when both are equal, false otherwise
     */
    equals(o: any): o is TwistPoint {
        if (!(o instanceof TwistPoint)) {
            return false;
        }

        const a = this.clone();
        a.makeAffine();

        const b = o.clone();
        b.makeAffine();

        return a.x.equals(b.x) && a.y.equals(b.y) && a.z.equals(b.z) && a.t.equals(b.t);
    }

    /**
     * Get the string representation of the point
     * @returns the string representation
     */
    toString(): string {
        const cpy = this.clone();
        cpy.makeAffine();

        return `(${cpy.x.toString()},${cpy.y.toString()},${cpy.z.toString()})`;
    }
}
